staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go

```diff
@@ -194,6 +204,84 @@ func TestTranslateVSphereCSIPVToInTree(t *testing.T) {
 			},
 			expErr: false,
 		},
+		{
+			name: "translate valid vSphere CSI PV with topology Node Affinity rules to vSphere CSI PV with topology labels",
+			csiPV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						CSI: &v1.CSIPersistentVolumeSource{
+							Driver:       VSphereDriverName,
+							VolumeHandle: "e4073a6d-642e-4dff-8f4a-b4e3a47c4bbd",
+							FSType:       "ext4",
+							VolumeAttributes: map[string]string{
+								paramStoragePolicyName:         "vSAN Default Storage Policy",
+								AttributeInitialVolumeFilepath: "[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk",
+							},
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+					NodeAffinity: &v1.VolumeNodeAffinity{
+						Required: &v1.NodeSelector{
+							NodeSelectorTerms: []v1.NodeSelectorTerm{
+								{
+									MatchExpressions: []v1.NodeSelectorRequirement{
+										{
+											Key:      "topology.csi.vmware.com/zone",
+											Operator: v1.NodeSelectorOpIn,
+											Values:   []string{"z1"},
+										},
+										{
+											Key:      "topology.csi.vmware.com/region",
+											Operator: v1.NodeSelectorOpIn,
+											Values:   []string{"r1"},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			intreePV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+					Labels: map[string]string{"topology.kubernetes.io/zone": "z1", "topology.kubernetes.io/region": "r1"},
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+							VolumePath: "[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk",
+							FSType:     "ext4",
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+					NodeAffinity: &v1.VolumeNodeAffinity{
```

Comment:
The In-tree PV will have node affinity terms too?

Reply:
I am not sure about it, but if at all we have node affinity rules present on the PV, we need to translate it so the schedular can use the node with topology keys applied by the vSphere CSI Driver.